### PR TITLE
e2e-test: support ApisixRoute v2

### PR DESCRIPTION
**Pull request by AlinsRan**
### Type of change:

- [x] Bugfix
- [ ] New feature provided
- [ ] Improve performance
- [ ] Backport patches

### What this PR does / why we need it:

fix: #1108 

### Pre-submission checklist:

* [ ] Did you explain what problem does this PR solve? Or what new features have been added?
* [ ] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [ ] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix-ingress-controller#community) first**

### CODE REVIEW

File: test/e2e/scaffold/scaffold.go [hunk 1]
```diff
@@ -583,11 +606,38 @@ func (s *Scaffold) CreateVersionedApisixPluginConfig(yml string) error {
...
+	}
+	if strings.Contains(yml, "kind: ApisixConsumer") {
+		ac := s.replaceApiVersion(yml, s.opts.APISIXConsumerVersion)
+		return s.CreateResourceFromString(ac)
+	}
+	if strings.Contains(yml, "kind: ApisixPluginConfig") {
+		apc := s.replaceApiVersion(yml, s.opts.ApisixPluginConfigVersion)
+		return s.CreateResourceFromString(apc)
 	}
+	return errors.New("this resource does not support")
```
Comment (lingsamuel) [inline test/e2e/scaffold/scaffold.go:622]: would be better to show what kind of resource it is. Use regex to extract the resource kind and return error in default branch of switch/case.

File: test/e2e/scaffold/scaffold.go [hunk 2]
```diff
@@ -583,11 +606,38 @@ func (s *Scaffold) CreateVersionedApisixPluginConfig(yml string) error {
...
+	}
+	if strings.Contains(yml, "kind: ApisixConsumer") {
+		ac := s.replaceApiVersion(yml, s.opts.APISIXConsumerVersion)
+		return s.CreateResourceFromStringWithNamespace(ac, namespace)
+	}
+	if strings.Contains(yml, "kind: ApisixPluginConfig") {
+		apc := s.replaceApiVersion(yml, s.opts.ApisixPluginConfigVersion)
+		return s.CreateResourceFromStringWithNamespace(apc, namespace)
+	}
+	return errors.New("this resource does not support")
```
Comment (lingsamuel) [inline test/e2e/scaffold/scaffold.go:638]: same

File: test/e2e/scaffold/scaffold.go (hunk 1 above)
Comment (tao12345666333) [reply test/e2e/scaffold/scaffold.go:622]: +1

Comment (AlinsRan) [reply test/e2e/scaffold/scaffold.go]: Updated.

File: test/e2e/scaffold/scaffold.go (hunk 2 above)
Comment (AlinsRan) [reply test/e2e/scaffold/scaffold.go]: Updated.

File: test/e2e/suite-ingress/resourcesync.go [hunk 3]
```diff
@@ -132,96 +123,120 @@ spec:
...
+	ginkgo.Describe("suite-ingress: scaffold v2beta3", func() {
+		suites(scaffold.NewScaffold(&scaffold.Options{
+			Name:                       "sync",
+			Kubeconfig:                 scaffold.GetKubeconfig(),
+			APISIXConfigPath:           "testdata/apisix-gw-config.yaml",
+			IngressAPISIXReplicas:      1,
+			HTTPBinServicePort:         80,
+			ApisixResourceVersion:      scaffold.ApisixResourceVersion().V2beta3,
+			ApisixResourceSyncInterval: "60s",
+		}))
```
Comment (tao12345666333) [inline test/e2e/suite-ingress/resourcesync.go:229]: ditto

File: test/e2e/suite-ingress/resourcesync.go [hunk 4]
```diff
@@ -132,96 +123,120 @@ spec:
...
+	ginkgo.Describe("suite-ingress: scaffold v2", func() {
+		suites(scaffold.NewScaffold(&scaffold.Options{
+			Name:                       "sync",
+			Kubeconfig:                 scaffold.GetKubeconfig(),
+			APISIXConfigPath:           "testdata/apisix-gw-config.yaml",
+			IngressAPISIXReplicas:      1,
+			HTTPBinServicePort:         80,
+			ApisixResourceVersion:      scaffold.ApisixResourceVersion().V2,
+			ApisixResourceSyncInterval: "60s",
+		}))
```
Comment (tao12345666333) [inline test/e2e/suite-ingress/resourcesync.go:240]: ditto

File: test/e2e/suite-ingress/webhook.go [hunk 5]
```diff
@@ -64,11 +54,35 @@ spec:
...
+	ginkgo.Describe("suite-ingress: scaffold v2", func() {
+		suites(scaffold.NewScaffold(&scaffold.Options{
+			Name:                  "webhook",
+			Kubeconfig:            scaffold.GetKubeconfig(),
+			APISIXConfigPath:      "testdata/apisix-gw-config.yaml",
+			IngressAPISIXReplicas: 1,
+			HTTPBinServicePort:    80,
+			ApisixResourceVersion: scaffold.ApisixResourceVersion().V2,
+			EnableWebhooks:        false,
+		}))
```
Comment (tao12345666333) [inline test/e2e/suite-ingress/webhook.go:86]: ditto

File: test/e2e/scaffold/scaffold.go [hunk 6]
```diff
@@ -583,11 +613,40 @@ func (s *Scaffold) CreateVersionedApisixPluginConfig(yml string) error {
...
+		return s.CreateResourceFromStringWithNamespace(ar, namespace)
+	case "ApisixConsumer":
+		ac := s.replaceApiVersion(yml, s.opts.APISIXConsumerVersion)
+		return s.CreateResourceFromStringWithNamespace(ac, namespace)
+	case "ApisixPluginConfig":
+		apc := s.replaceApiVersion(yml, s.opts.ApisixPluginConfigVersion)
+		return s.CreateResourceFromStringWithNamespace(apc, namespace)
+	}
+	errString := fmt.Sprint("the resource ", kindValue, " does not support")
+	return errors.New(errString)
```
Comment (lingsamuel) [inline test/e2e/scaffold/scaffold.go:647]: `fmt.Errorf`

File: test/e2e/scaffold/scaffold.go [hunk 7]
```diff
@@ -541,12 +564,19 @@ func (s *Scaffold) FormatNamespaceLabel(label string) string {
...
 	versionRegex = regexp.MustCompile(`apiVersion: apisix.apache.org/v.*?\n`)
+	kindRegex    = regexp.MustCompile(`kind: .*?\n`)
 )
 
 func (s *Scaffold) replaceApiVersion(yml, ver string) string {
 	return versionRegex.ReplaceAllString(yml, "apiVersion: "+ver+"\n")
 }
 
+func (s *Scaffold) getKindValue(yml string) string {
+	kind := strings.Replace(kindRegex.FindString(yml), "\n", "", -1)
```
Comment (lingsamuel) [inline test/e2e/scaffold/scaffold.go:575]: Why not use capture groups?

File: test/e2e/suite-endpoints/endpoints.go [hunk 8]
```diff
@@ -119,27 +110,48 @@ spec:
...
+	ginkgo.Describe("suite-endpoints: scaffold v2beta3", func() {
+		suites(scaffold.NewScaffold(&scaffold.Options{
+			Name:                  "endpoints-port",
+			Kubeconfig:            scaffold.GetKubeconfig(),
+			APISIXConfigPath:      "testdata/apisix-gw-config.yaml",
+			IngressAPISIXReplicas: 1,
+			HTTPBinServicePort:    8080,
+			ApisixResourceVersion: scaffold.ApisixResourceVersion().V2beta3,
+		}))
+	})
```
[the remark this answers is not in the document]
Comment (AlinsRan) [reply test/e2e/suite-endpoints/endpoints.go:146]: > how ablout using `suites(scaffold.NewDefaultScaffold())` ?

Separate configuration item `HTTPBinServicePort:8080` are used here.

File: test/e2e/suite-endpoints/endpoints.go [hunk 9]
```diff
@@ -119,27 +110,48 @@ spec:
...
+	ginkgo.Describe("suite-endpoints: scaffold v2", func() {
+		suites(scaffold.NewScaffold(&scaffold.Options{
+			Name:                  "endpoints-port",
+			Kubeconfig:            scaffold.GetKubeconfig(),
+			APISIXConfigPath:      "testdata/apisix-gw-config.yaml",
+			IngressAPISIXReplicas: 1,
+			HTTPBinServicePort:    8080,
+			ApisixResourceVersion: scaffold.ApisixResourceVersion().V2,
+		}))
```
[the remark this answers is not in the document]
Comment (AlinsRan) [reply test/e2e/suite-endpoints/endpoints.go:155]: Separate configuration item HTTPBinServicePort:8080 are used here.

File: test/e2e/suite-ingress/resourcesync.go (hunk 3 above)
Comment (AlinsRan) [reply test/e2e/suite-ingress/resourcesync.go:229]: Separate configuration item `ApisixResourceSyncInterval: "60s"` are used here.

File: test/e2e/suite-ingress/resourcesync.go (hunk 4 above)
Comment (AlinsRan) [reply test/e2e/suite-ingress/resourcesync.go:240]: Separate configuration item `ApisixResourceSyncInterval: "60s"` are used here.

File: test/e2e/suite-ingress/webhook.go (hunk 5 above)
Comment (AlinsRan) [reply test/e2e/suite-ingress/webhook.go:86]: This is to identify `enablewebhooks`.

**Review by tao12345666333 — APPROVED** [review]
This PR is very big, basically LGTM

Let's merge first (I like the split of CI jobs in this PR, it can greatly reduce the time consuming of CI)

File: test/e2e/scaffold/scaffold.go (hunk 7 above)
Comment (AlinsRan) [reply test/e2e/scaffold/scaffold.go:575]: Didn't notice.

Comment (tao12345666333) [reply test/e2e/scaffold/scaffold.go:575]: How about a fix in a follow-up PR? @AlinsRan

**Review by lingsamuel — APPROVED** [review]

File: test/e2e/scaffold/scaffold.go (hunk 7 above)
Comment (AlinsRan) [reply test/e2e/scaffold/scaffold.go:575]: > How about a fix in a follow-up PR? @AlinsRan

Sure.

**Comment by tao12345666333** [comment]
let's move forward. thanks

**Pull request merged**